volumina/utility/preferences.py

```diff
@@ -25,75 +25,89 @@
         self.setmany((group, setting, value))
 
-    def getmany(self, *args: Tuple[str, str, Any]) -> Tuple[Any, ...]:
+    def getmany(self, *args: Tuple[str, str, Any]) -> Tuple:
         with self._lock:
+            data = self.read()
             result = []
             for group, setting, default in args:
                 try:
-                    result.append(self._data[group][setting])
+                    result.append(data[group][setting])
                 except KeyError:
                     result.append(default)
             return tuple(result)
 
     def setmany(self, *args: Tuple[str, str, Any]) -> None:
         with self._lock:
+            data = self.read()
             for group, setting, value in args:
-                self._data.setdefault(group, {})[setting] = value
-            try:
-                with open(self._path, "wb") as f:
-                    # Use the lowest available protocol for backwards compatibility.
-                    pickle.dump(self._data, f, protocol=0)
-            except Exception as e:
-                logger.exception(f"Failed to save preferences to {str(self._path)!r}")
-                if not isinstance(e, IOError):
-                    self._path.unlink()
-
-    def get_location(self) -> pathlib.Path:
-        with self._lock:
-            return self._path
+                data.setdefault(group, {})[setting] = value
+            self.write(data)
 
-    def set_location(self, path: os.PathLike) -> None:
-        with self._lock:
-            self._path = pathlib.Path(path)
-            self._data = _load_preferences(self._path)
+    def read(self):
+        try:
+            with open(self.path) as f:
+                return json.load(f)
+        except FileNotFoundError:
+            return {}
+
+    def write(self, data):
+        os.makedirs(self.path.parent, exist_ok=True)
+        with open(self.path, "w") as f:
+            json.dump(data, f, indent=2, sort_keys=True)
 
+    def migrate(self, old_path):
+        if not old_path.exists():
+            return
 
-def _load_preferences(path: pathlib.Path) -> MutableMapping[str, MutableMapping[str, Any]]:
-    try:
-        with open(path, "rb") as f:
-            return pickle.load(f)
-    except FileNotFoundError:
-        return {}
-    except Exception as e:
-        logger.exception(f"Failed to load preferences from {str(path)!r}")
-        if not isinstance(e, IOError):
-            path.unlink()
-        return {}
+        import pickle
 
+        try:
+            with open(old_path, "rb") as f:
+                old_preferences = pickle.load(f)
+        except Exception:
+            return
 
-_preferences = _Preferences(pathlib.Path.home() / ".ilastik_preferences")
+        old_path.unlink()
+        self.write(old_preferences)
+
+
+_preferences = Preferences(
+    os.path.join(
+        appdirs.user_config_dir(appname="ilastik", appauthor=False),
+        "preferences.json",
+    )
+)
```

Comment:
Seems like this would be a good place to have the several Paths where the config file could reside if we were to allow for that kind of thing

Reply:
I don't think we need multiple locations for `preferences.json` since is is created automatically by volumina, and users won't see it under normal circumstances.